client/shared/src/test/scala/org/http4s/client/ClientSuite.scala

```diff
@@ -147,4 +148,42 @@ class ClientSpec extends Http4sSuite with Http4sDsl[IO] {
       .attempt
       .assertEquals(Left(MyThrowable), "Throwable did not get handled as expected")
   }
+
+  test("mock client should drain the body if it has not been consumed") {
+
+    def app(finalized: Ref[IO, Boolean]) = HttpApp[IO] { (_: Request[IO]) =>
+      Response[IO]()
+        .pipeBodyThrough(_.onFinalize(finalized.set(true)))
+        .pure[IO]
+    }
+
+    for {
+      finalized <- Ref.of[IO, Boolean](false)
+      client = Client.fromHttpApp(app(finalized))
+      _ <- client.status(Request[IO]()).void
+      result <- finalized.get
+    } yield assertEquals(result, true)
+  }
+
+  test("mock client should drain the body if it has been partially") {
```

Comment:
Besides testing that the body is drained, I think we also want to test that the stream is not compiled more than once.